### ISSUE TITLE
place 'maven-wrapper.jar' in ${user.home}/.m2/wrapper

### ISSUE THREAD

**Issue by gzm55** [takari/maven-wrapper#77]
Hi, now the layout of maven-wrapper is : 
```
${basedir}/mvnw
${basedir}/mvnw.cmd
${basedir}/.mvn/wrapper/maven-wrapper.jar
${basedir}/.mvn/wrapper/maven-wrapper.properties
${basedir}/.mvn/wrapper/MavenWrapperDownloader.java

${user.home}/.m2/wrapper/dist/xxx/apache-maven-<ver>-bin/
```

The downloadable artifact `maven-wrapper.jar` is placed in the project dir. Can we cache this jar in user conf dir at `${user.home}/.m2/wrapper/maven-wrapper-<ver>.jar` or `${user.home}/.m2/wrapper/launcher/maven-wrapper-<ver>.jar`? Then there is no generated files of the building project in non-`${basedir}/target` dir.

**Comment by gmussi** [comment]
I will also vouch for this. It is cleaner this way.

**Comment by lukaszlenart** [comment]
Same here 👍

**Comment by mosabua** [comment]
The reason the jar is in the project is that the whole idea of Maven wrapper is to be able to check out a source repository and run the desired mvnw command without any other dependencies than an installed JDK. Relying on the existence of a .m2/repository/... is therefore not possible and misses the whole point of the project.

If you dont want to have the wrapper jar in place we already have support for this with the latest release that can download the jar via curl or wget or even compile a downloader and get it.

It might be worth considering adding the local maven repo as an additional path to look in but to be honest there are very minimal gains to be had from implementing this imho. Feel free to check it out and see if you want to send a PR.

**Comment by lukaszlenart** [comment]
You relay on existence of the jar anyway, in this or other folder it doesn't matter. If you are using wrapper in multiple projects, each  time the wrapper must download exactly the same jar and keep it locally.

**Comment by lukaszlenart** [comment]
There is one more use case where placing wrapper in a common place is a good idea - when you have a multimodule project like [Struts](https://github.com/apache/struts/). Very often I build submodules to test something. With current setup I have to duplicated exactly the same wrapper in each submodule.

**Comment by mosabua** [comment]
No you dont. The wrapper just has to be in the top level where you run your build. If you want to built a submodule you just use the -pl option and are fine. But as I mentioned .. if you have a PR that improves the current impl .. I am ready to look.

**Comment by lukaszlenart** [comment]
Thanks for the tip!

**Comment by seanf** [comment]
At least in projects which don't keep the jar in source control (https://github.com/takari/maven-wrapper/issues/24) it would be nice if the wrapper would put the jar in `${user.home}/.m2/wrapper`. 

For one thing, a central location would be a little easier to cache with say, Travis's caching feature. Or for local (non-CI) development, it would greatly improve the chances of finding the jar, since the developer may well have it from building another mvnw project.

**Comment by awittha** [comment]
I'm concerned about keeping the JAR _in the user home_ when not kept in the project. For specific workflows, it may be beneficial but I can think of a couple situations where a single, user-wide JAR would be problematic:

1. Building 2 projects at once that are using different, incompatible versions of maven-wrapper (such as if one is old and not updated) - the JAR is not uniquely-named. One project will think it has found a compatible maven-wrapper JAR and will be wrong.
2. Building 2 projects at once, with no previously-downloaded maven-wrapper: which process gets to lock the file and write the JAR? Does the other one error?

It's _always_ safe - though maybe not optimal - to keep a one maven-wrapper JAR per project, in the project directory. It's _not_ necessarily always safe to keep a single user-wide JAR.

Moreover, wouldn't it make more sense for the central location of the maven-wrapper JAR to be in `~/.m2/repository/com/takari/....`?

**Comment by seanf** [comment]
> 1. Building 2 projects at once that are using different, incompatible versions of maven-wrapper (such as if one is old and not updated) - the JAR is not uniquely-named. One project will think it has found a compatible maven-wrapper JAR and will be wrong.

True, depending on the compatibility policy. This might be handled by including version in the filename, especially if it goes under `~/.m2/repository/io/takari/maven-wrapper/`

> 2. Building 2 projects at once, with no previously-downloaded maven-wrapper: which process gets to lock the file and write the JAR? Does the other one error?

Locking: yes, good point. But now I have to wonder about the locking of Maven jar files under `${user.home}/.m2/wrapper/dist/xxx/apache-maven-<ver>-bin/`... Doesn't look like there is any: https://github.com/takari/maven-wrapper/blob/fe0fa453699b323ec1ee40ccc61794723458c6ae/src/main/java/org/apache/maven/wrapper/DefaultDownloader.java#L54-L69

**Comment by gzm55** [comment]
1. a versioned pattern of jar path works.
2. lock a file in a portable way is not an easy work, but we can first download (if not already downloaded) the wrapper jar to a tmp file, then move it to the dest path. this is a safe way even when replacing a jar loaded by a running java process. And as @seanf said, the way of downloading and unzipping the maven dist jar within the wrapper itself is more unsafe in concurrent bootstrap workflows.

3. putting wrapper in `~/.m2/repository/com/takari/` is not ok, because `~/.m2/repository` can be re-defined to another path by user/global settings.xml, and it is hard to parse these files in a simple shell script.

**Comment by awittha** [comment]
> Locking: yes, good point. But now I have to wonder about the locking of Maven jar files under `${user.home}/.m2/wrapper/dist/xxx/apache-maven-<ver>-bin/...` Doesn't look like there is any:

You are correct and this was an issue for me in CI systems; I had to add bootstrap logic to each build to keep `~/.m2/wrapper` unique-per-build (without making `~/.m2/repository` unique-per-build) to avoid issues.

_(Takari's done work on [solving this for _maven-managed_ downloads](http://takari.io/book/30-team-maven.html))_

**Comment by seanf** [comment]
@gzm55 said:
> we can first download (if not already downloaded) the wrapper jar to a tmp file, then move it to the dest path. this is a safe way even when **replacing a jar loaded by a running java process**.

I don't think that will work on Windows though.

@awittha said:
> (Takari's done work on [solving this for maven-managed downloads](http://takari.io/book/30-team-maven.html))

You mean the Takari Concurrent Local Repository? Watch out for these issues: https://github.com/takari/takari-local-repository/issues/4 and https://github.com/takari/takari-local-repository/issues/5

I ended up disabling it in my projects and using `-Daether.connector.resumeDownloads=false` as [suggested here](https://github.com/takari/takari-local-repository/issues/5#issuecomment-264161273).

**Comment by gzm55** [comment]
@seanf when bootstrap the wrapper jar and even the maven dist package, project- and home- based dirs both suffer the race condition of multiple build instances. Since this bootstrap process must be portable, lightweight, and lock-free, we can limit the race as much as possible, and the worst case is fail fast with a successful re-building. So imo the outline of the process is:

1. the result path must be uniq by content (via hash and version) and central (for re-use the result).
2. for each bootstrap instance, create a uniq temporary runtime dir. This step is not very portable in windows, but if using timestamp+processId+random+sequenceNum, we almost never encounter the race condition.
3. download and unzip, already done.
4. before move, test again the existence of dest path, then the race condition is limited to the filesystem operations of checking existence and moving files.
5. move the downloaded artifacts to the dest path (one file or one directory), then
  - on a fs with atomic `mv`
    - mv success: just begin the following works
    - mv fail: skip fail, just begin the following works, since the contents with the same path is always the same
  - on a fs without atomic `mv`
    - mv success for the very first one but the following works are interrupted by the other mv action
    - mv success by replacing the existence one
    - mv fail
6. for every result in step 5, if the process fails any way, we can always repeat (for limited times) steps 4 and 5, since the dest content of mv operation is idempotent, the existing of dest path means success of the bootstrap.

The steps 4-6 is lock-free, when all re-tries fail, we still have to fail the whole bootstrap. I think it is a better action than producing any interrupted results.

**Comment by mosabua** [comment]
You can already do that now by setting a URL for the wrapper jar in maven-wrapper.properties.  it might even support a file URL .. but I have not tried. In either case .. the jar does not have to be in the local repo folder in the version control system at this stage. If you want any improvements implemented please send a PR.
